src/syscore/windows/iocp/iocp.rs

```diff
@@ -0,0 +1,352 @@
+use winapi::um::handleapi::INVALID_HANDLE_VALUE;
+use winapi::um::ioapiset;
+use winapi::um::minwinbase::OVERLAPPED_ENTRY;
+use winapi::um::winbase::INFINITE;
+use winapi::um::winnt::HANDLE;
+
+pub struct SysProactor {
+    iocp_handle: WinHandle,
+    completions: HashMap<WinHandle, oneshot::Sender<usize>>,
+    // ULONG_PTR
+    completion_key: AtomicUsize,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct WinOverlappedEntry(OVERLAPPED_ENTRY);
+
+impl std::ops::Deref for WinOverlappedEntry {
+    type Target = OVERLAPPED_ENTRY;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl From<OVERLAPPED_ENTRY> for WinOverlappedEntry {
+    fn from(h: OVERLAPPED_ENTRY) -> Self {
+        Self(h)
+    }
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct WinHandle(HANDLE);
+
+// TODO [igni]: SUPER CURSED THIS NEEDS TO GO AWAY
+unsafe impl Send for WinHandle {}
+unsafe impl Sync for WinHandle {}
+// TODO [igni]: SUPER CURSED THIS NEEDS TO GO AWAY
+unsafe impl Send for WinRawHandle {}
+unsafe impl Sync for WinRawHandle {}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct WinRawHandle(RawHandle);
+
+impl std::ops::Deref for WinHandle {
+    type Target = HANDLE;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl From<HANDLE> for WinHandle {
+    fn from(h: HANDLE) -> Self {
+        Self(h)
+    }
+}
+
+impl std::ops::Deref for WinRawHandle {
+    type Target = RawHandle;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl From<RawHandle> for WinRawHandle {
+    fn from(h: RawHandle) -> Self {
+        Self(h)
+    }
+}
+
+impl SysProactor {
+    /// Returns a reference to the proactor.
+    pub fn new() -> Result<Self, ()> {
+        let completion_key = AtomicUsize::default();
+        let iocp_handle = try_create_iocp(
+            WinRawHandle(INVALID_HANDLE_VALUE as _),
+            WinHandle(ptr::null_mut()),
+            completion_key.fetch_add(1, Ordering::SeqCst),
+        )
+        .unwrap();
+
+        std::thread::spawn(|| loop {});
+
+        Ok(Self {
+            iocp_handle,
+            completion_key,
+            completions: Default::default(),
+        })
+    }
+
+    /// Wakes the thread waiting on proactor.
+    pub fn wake(&self) -> io::Result<()> {
+        Ok(())
+    }
+
+    /// Wait for completion of IO object
+    pub fn wait(&self, max_event_size: usize, duration: Option<Duration>) -> io::Result<usize> {
+        // Wait for completion
+        let mut bytes_transferred = 0;
+        let mut actual_completion_key = 0;
+        let mut entries_removed = ptr::null_mut();
+
+        let mut entries_stuff = &mut self.completions.iter().map(|stuff| stuff.key());
+
+        let done = unsafe {
+            ioapiset::GetQueuedCompletionStatusEx(
+                *handle,
+                entries_stuff,
+                &mut stuff.key(),
+                entries_removed,
+                1, // TODO: Play around with that on poll or something.
```

Comment:
This lacks a last parameter: fAlertable (https://docs.microsoft.com/en-us/windows/win32/fileio/getqueuedcompletionstatusex-func)